matchzoo/models/anmm_model.py

```diff
@@ -0,0 +1,68 @@
+"""An implementation of aNMM Model."""
+
+
+import logging
+import keras
+
+from keras.activations import softmax
+from keras.initializers import RandomUniform
+
+from matchzoo import engine
+
+logger = logging.getLogger(__name__)
+
+
+class ANMMModel(engine.BaseModel):
+    """
+    ANMM Model.
+
+    Examples:
+        >>> model = ANMMModel()
+        >>> model.guess_and_fill_missing_params(verbose=0)
+        >>> model.build()
+
+    """
+
+    @classmethod
+    def get_default_params(cls) -> engine.ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params(with_embedding=True)
+        params.add(engine.Param('bin_num', 60))
+        params.add(engine.Param('dropout_rate', 0.1))
+        params.add(engine.Param('left_text_len', 10))
+        params.add(engine.Param('num_layers', 2))
+        params.add(engine.Param('hidden_sizes', [10, 10]))
+        return params
+
+    def build(self):
+        """
+        Build model structure.
+
+        aNMM model based on bin weighting and query term attentions
+        """
+        # query is [batch_size, left_text_len]
+        # doc is [batch_size, right_text_len, bin_num]
+        query, doc = self._make_inputs()
+        embedding = self._make_embedding_layer()
+        initializer_fc = RandomUniform(minval=-0.1, maxval=0.1, seed=11)
+        initializer_gate = RandomUniform(minval=-0.01, maxval=0.01, seed=11)
+
+        q_embed = embedding(query)
+        q_w = keras.layers.Dense(
+            1, kernel_initializer=initializer_gate, use_bias=False)(q_embed)
+        q_w = keras.layers.Lambda(
+            lambda x: softmax(x, axis=1),
+            output_shape=(self._params['left_text_len'],)
+        )(q_w)
+        z = doc
+        z = keras.layers.Dropout(rate=self._params['dropout_rate'])(z)
+        for i in range(self._params['num_layers'] - 1):
```

Comment:
For the dense layers, you can also refer to [DRMM](https://github.com/NTMC-Community/MatchZoo/blob/feature/add_drmm/matchzoo/models/drmm.py) to simplify your code, since the dense layer has been defined in the `base_model` just like `embedding`.

Reply:
As we discussed, this will decrease the flexibility of the MLP layers. I will update this later if we want to support tuning hyper-parameters with hyper-opt.

Reply:
The current implementations of models like DUET, DSSM,CDSSM, etc also use Dense layers to implement MLP. If we update all these models, the cost would be quite high and it also adds limitations on the supported MLP architectures (e.g. all activation functions and the number of nodes in each layer must be the same). We can think of other solutions if we want to support tuning hyper-parameters with hyper-opt.